### DISCUSSION TITLE
Error attempting to train TextCat Classifier

### DISCUSSION THREAD

**Discussion by morankyle** [explosion/spaCy#8100] [Help: Coding & Implementations] [answered]
When running "python -m spacy train config.cfg --output ./output" I get the error "Connection error, and we cannot find the requested files in the cached path. Please try again or make sure your internet connection is on"

The computer I am attempting to make this model on does not have internet connectivity however I can use pip  to install certain packages like spacy-transformers. 

Does this error mean I need to be connected to the internet to train this model? Or did I just make a mistake setting the paths for the data files?

**Comment by svlandeg** [comment, marked as answer]
This error is being thrown because you're trying to use a Huggingface transformer model in your `nlp` pipeline. This model is specified by name in the config file, and is being downloaded from HuggingFace when you run the training. And then this happens: https://huggingface.co/transformers/master/troubleshooting.html#firewalled-environments

Instead of providing the name of the model, you could also download the model directory once, store it on your machine, and refer to that path from within the config file instead:
```
[transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "your_cached_directory"
```
After that, I think/hope you shouldn't have a problem training on an offline machine.

**Reply by morankyle** [reply]
Thank you so much! Will give it a try now

**Reply by morankyle** [reply]
It worked! Follow up question, would this be the correct format for multiple classes for text cat?

def make_docs(data);
docs = []
a = tqdm(nlp.pipe(data, as_tuples=True), total=len(data))
for doc, label in tqdm(nlp.pipe((data), as_tuples=True), total = len(data)):
if label == 'F':
doc.cats['F'] = 1
doc.cats['D'] = 0
doc.cats['E'] = 0
elif label == 'D':
doc.cats['F'] = 0
doc.cats['D'] = 1
doc.cats['E'] = 0
elif label == 'E':
doc.cats['F'] = 0
doc.cats['D'] = 0
doc.cats['E'] = 1

docs.append(doc)
return docs

**Reply by morankyle** [reply]
As I'm currently getting an error that labels for textcat are not initialized

**Reply by svlandeg** [reply]
How exactly did you create the `textcat` component, and did you call [`initialize`](https://spacy.io/api/textcategorizer#initialize)? You need to call `initialize` to make sure the component is ready for training. You can give it a data sample with the labels correctly annoted (as in your snippet), and then the `textcat` will deduce those labels automatically. Alternatively, you can give an argument `labels` to the `initialize` call. Finally, you can also call `add_label` specifically, though you usually don't need that if you've called `initialize` properly.

Some more background information can be found here: https://spacy.io/usage/training#initialization

**Reply by morankyle** [reply]
[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatBow.v1"
exclusive_classes = True
ngram_size = 1
no_output_layer = false
n0 = null


The initialize step is definitely where I messed up as I have this currently:

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

**Reply by morankyle** [reply]
[initialize.components.textcat]

[initialize.components.textcat.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/textcat.json"

I understand that I could add this to my config,cfg and it should fix the problem, only question is in regards to the following CLI command, what should the format of the python file look like?

python -m spacy init labels config_path output_path --code --verbose --gpu-id overrides

**Reply by svlandeg** [reply]
I'm not sure I understand which "python file" you're referring to in your last question.

In general though, if you're using the train CLI with a config generated by `spacy init config`, and feed the `spacy train` command appropriate data, you won't have to worry about the initialization. Behind the scenes, the training data will be used to deduce the labels.

In short, I wonder whether your data is being read correctly?

Can you take two steps back and provide us with the full config file _and_ CLI command for training that you've used, where you got the error about the textcat not being initialized, _and_ can you copy/paste the full log of the command, including the full error stack trace?

I'm afraid that otherwise we're left guessing what exactly might be causing you these problems.

**Reply by morankyle** [reply]
Below is both the code to create the docs and the CLI commands, the actual code is behind firewall at work so tomorrow when back in office I can copy the exact error / log of the command. Thank you so much




**Create Docs:**

docs = []

a = tqdm(nlp.pipe(data, as_tuples=True), total = len(data))

for doc, label in tqdm(nlp.pipe(data, as_tuples=True), total = len(data)):
if label == 'small':
doc.cats["small"] = 1
doc.cats["medium"] = 0
doc.cats["big"] = 0
elif label == 'medium':
doc.cats["small"] = 0
doc.cats["medium"] = 1
doc.cats["big"] = 0
elif label == 'big':
doc.cats["small"] = 0
doc.cats["medium"] = 0
doc.cats["big"] = 1


docs.append(doc)

return docs



**CLI:**

python -m spacy init fill-config ./base_config.cfg ./config.cfg

python -m spacy train config.cfg --output ./output

**Reply by svlandeg** [reply]
Thanks! And what is the output when you run 
```
python -m spacy debug data config.cfg
```
?

**Reply by morankyle** [reply]
Okay tomorrow first thing I will post that error message as well, again thank you so much and I'm sure Im just making a silly error somewhere

**Reply by svlandeg** [reply]
No worries, we'll get to the bottom of it :-)

**Reply by morankyle** [reply]
**CLI Command to train model:**

python -m spacy train config.cfg --output ./output

**Error after that command:**

ValueError: [E143] Labels for component 'textcat' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's 'initialize' method. 

**Error after running debug command:** 

Corpus is loadable 

ValueError: [E143] Labels for component 'textcat' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's 'initialize' method.

**Reply by morankyle** [reply]
**Full Config File:**



[paths]
train = "data/train.spacy"
dev = "data/valid.spacy"
init_tok2vec = null
vectors = null 
roberta = "roberta_base"

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat"]
batch_size = 128
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null

[components]

[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = true
ngram_size = 1
no_output_layer = false
nO = null

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = ${paths.roberta}

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer.model.tokenizer_config]
use_fast = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 500
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
cats_score = 1.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

**Reply by svlandeg** [reply]
> Corpus is loadable
ValueError: [E143] Labels for component 'textcat' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's 'initialize' method.

This does imply that the training examples that you've got stored in your `.spacy` files, don't actually contain the right categories.

What do you get when you do this?
```
doc_bin = DocBin().from_disk("data/train.spacy")
docs = list(doc_bin.get_docs(nlp.vocab))
print(len(docs))
for doc in docs:
    print(doc.cats)
```
Can you confirm the `cats` are defined correctly in those `docs`?

**Reply by morankyle** [reply]
**Output:**

1 
{'F': 0, 'D': 0, 'M': 0, 'E': 0, 'N': 0, 'I': 1}

**Reply by svlandeg** [reply]
Hm, odd, it does look like your training file contains the correct categories, and the `textcat` should be using those to initialize itself. 

One more idea: can you try increasing the `max_length` parameter as described [here](https://github.com/explosion/spaCy/discussions/8128#discussioncomment-754044) (or just set to 0)?

**Comment by filipematos95** [comment]
> Instead of providing the name of the model, you could also download the model directory once, store it on your machine, and refer to that path from within the config file 

@svlandeg  How can we download the right model to use it offline?

**Reply by svlandeg** [reply]
> @svlandeg How can we download the right model to use it offline?

You can find the model directories on https://huggingface.co/models. Basically you'd download an entire directory and provide the path to the directory as the `name` when creating a `TransformerModel` with `spacy_transformers`. This `str_path` will be used to call 
```
from transformers import AutoModel, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained(str_path, **tok_config)
transformer = AutoModel.from_pretrained(str_path)
```
which will initialize the model with the pretrained weights from Huggingface.
